core/src/main/java/amino/run/kernel/client/KernelClient.java

```
@@ -28,41 +34,180 @@
+            servers.put(host, serverInfo);
+            metrics.put(host, metric);
+            return serverInfo;
         } catch (Exception e) {
-            logger.severe("Could not find MicroService server on host: " + e.toString());
+            logger.severe(
+                    String.format(
+                            "Could not find kernel server on host: %s. Exception: %s", host, e));
         }
         return null;
     }
 
+    /**
+     * Removes cached remote kernel server information for the given host, if it has missed
+     * heartbeats for {@link amino.run.kernel.client.KernelClient#MAX_HEARTBEAT_MISS_TIMES} times
+     *
+     * @param host
+     */
+    private void removeHost(InetSocketAddress host) {
+        KernelServerInfo server = servers.get(host);
+        if (server != null) {
+            if ((server.refCount - server.refCountOnHeartBeat >= MAX_HEARTBEAT_MISS_TIMES)) {
+                servers.remove(host);
+                metrics.remove(host);
+            }
+        }
+    }
+
+    /**
+     * Gets the remote kernel server information for the given host
+     *
+     * @param host
+     * @return Kernel Server Information
+     */
+    private KernelServerInfo getServerInfo(InetSocketAddress host) {
+        KernelServerInfo serverInfo = servers.get(host);
+        if (serverInfo == null) {
+            serverInfo = addHost(host);
+        }
+        return serverInfo;
+    }
+
+    /**
+     * Gets the RMI stub reference to remote kernel server for the given host
+     *
+     * @param host
+     * @return RMI stub to remote kernel server
+     */
     private KernelServer getServer(InetSocketAddress host) {
-        KernelServer server = servers.get(host);
-        if (server == null) {
-            server = addHost(host);
+        KernelServerInfo serverInfo = getServerInfo(host);
+        return serverInfo != null ? serverInfo.remoteRef : null;
```

I don't like passing nulls around like this.  People forget to check return values for null and then we get intermittent memory problems, rather than easier-to-debug exceptions.  Rather throw an exception?

---

PS: Because getServerInfo() never returns null, this method should also never need to return null.  If getServerInfo() returns null then we have a problem and throwing an exception is appropriate.

---

Agreed. Throwing `KernelServerNotFoundException` exception from `addHost()` upon failure and propagating it to `getServerInfo()` and `getServer()`.